### ISSUE TITLE
direct communication between razor and js in razor files

### ISSUE THREAD

**Issue by sajjadarashhh** [dotnet/aspnetcore#49582]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

we can create feature for communicate between js and c# directly in razor pages with some senarios that make easier to coding for developers and thats not effect performance because that can be handled in compile time

### Describe the solution you'd like

<script>
@size = document.getElementById('test').width;
</script>
@code{
  public int size{get;set}
}
js compiled to this:
<script>
DotNetRef.size=document.getElementById('test').width;
</script>

**Comment by sajjadarashhh** [comment]
I have another idea for that
We can create new keyword named: script
And write c# in and js in that like this:
```
@script{
  @size = document.getElementById(size).innerText:
}
```

**Comment by sajjadarashhh** [comment]
@danroth27 can you help me for this?

**Comment by marinasundstrom** [comment]
@sajjadarashhh I don't think that this is in line with what Razor is about. You are suggesting adding JavaScript-specific syntactic sugar into the Razor templating language. 

There are things that seriously would make this a bad and confusing experience in that it makes you mix two languages, C#/Razor/.NET and JavaScript, that don't execute in the same domain.

Then we have the fact that JavaScript can't execute everywhere, like when being pre-rendered on the server. So your code would possible throw in certain cases.

**Comment by sajjadarashhh** [comment]
@marinasundstrom i dont said we must be take that in this way
I just have an idea to make a way to use js and .net in new format that make it easier to use c#-js interoperability we can choose many ways for this something like new files depend on components that do this for us 
With new format something like jasor and write js functions that have access to c# objects in that component and use that functions in component
For many use cases developers want to configure js components easier with out write many codes for interop
And we can send blazor components to js libraries without writing many codes just with c# element refrences
This take razor more easier and developer friendly

**Comment by marinasundstrom** [comment]
@sajjadarashhh JS and C# each have some concepts that don't work well together. And blending the languages would arguably just make it more confusing for developers. That is why it has failed so many times. So we have to respect the boundaries between the different environments.

C# interoperability with JavaScript in Blazor is already quite good. In Blazor WebAssembly you can use source generators that generate C#/JS interop code for you. It is similar to how P/Invoke does it with native code.

https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-7.0

I used to think that proposal like these were great - so I sympathize. Nowadays, I realize that interoperability is far more complex.

**Comment by sajjadarashhh** [comment]
The goals of this idea is not thing we create a way to work c# and js working together
The goal of this is a create a way to make interop codes at compile and automatically
Each thing interop doing this pattern can do that
And this is not very complex as you discussed
#marinasundstrom

**Comment by marinasundstrom** [comment]
@sajjadarashhh I think that this would essentially be a new a language in itself. Trying to blur the lines between JS and C#.

Razor is basically a templating language on top of C#/.NET. Not an interop tool.

**Comment by javiercn** [comment]
This is not something that we are planning to implement in the near future.

The proposal ignores all the challenges involved in the communication between .NET and JavaScript around scheduling and re-rendering and has a high cost in the form of an additional dotnet -> JavaScript compiler.

Pushing the attributes from the JS side transparently to the server is not realistic and will result in unexpected outcomes in many common situations (like during prerendering).

The .NET side needs to be in control of when DOM attributes are accessed and do so only during OnAfterRender which is when it knows that the values are correct.
